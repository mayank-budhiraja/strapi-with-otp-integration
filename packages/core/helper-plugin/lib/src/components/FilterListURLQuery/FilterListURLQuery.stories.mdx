<!--- FilterListURLQuery.stories.mdx --->

import { useEffect } from 'react';
import { Meta, ArgsTable, Canvas, Story } from '@storybook/addon-docs';
import { Button } from '@strapi/design-system/Button';
import { Box } from '@strapi/design-system/Box';
import { Main } from '@strapi/design-system/Main';
import { Flex } from '@strapi/design-system/Flex';
import { Stack } from '@strapi/design-system/Stack';
import useQueryParams from '../../hooks/useQueryParams';
import FilterListURLQuery from './index';

<Meta title="components/FilterListURLQuery" />

# FilterListURLQuery

This component is used to change the page size.

## Imports

```js
import { FilterListURLQuery } from '@strapi/helper-plugin';
```

## Usage

<Canvas>
  <Story name="base">
    {() => {
      const [{ query }, setQuery] = useQueryParams();
      const filters = {
        $and: [{ city: { $eq: 'paris' } }, { categories: { name: { $eq: 'paris' } } }],
      };
      useEffect(() => {
        setQuery({ filters });
      }, []);
      const handleResetFilters = () => setQuery({ filters });
      return (
        <Main>
          <Stack spacing={6}>
            <Flex paddingTop={6}>
              <FilterListURLQuery
                filtersSchema={[
                  {
                    name: 'categories',
                    fieldSchema: {
                      type: 'relation',
                      mainField: { name: 'name', schema: { type: 'string' } },
                    },
                    metadatas: { label: 'categories' },
                  },
                  {
                    name: 'city',
                    fieldSchema: {
                      type: 'string',
                    },
                    metadatas: { label: 'city' },
                  },
                ]}
              />
            </Flex>
            <Box>
              <Button onClick={handleResetFilters}>Reset filters</Button>
            </Box>
          </Stack>
        </Main>
      );
    }}
  </Story>
</Canvas>

### Props

<ArgsTable of={FilterListURLQuery} />
