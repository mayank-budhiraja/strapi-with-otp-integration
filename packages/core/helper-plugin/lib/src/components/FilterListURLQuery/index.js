/**
 *
 * FilterListURLQuery
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import useQueryParams from '../../hooks/useQueryParams';
import AttributeTag from './AttributeTag';

const FilterListURLQuery = ({ filtersSchema }) => {
  const [{ query }, setQuery] = useQueryParams();

  const handleClick = filter => {
    const nextFilters = query.filters.$and.filter(prevFilter => {
      const name = Object.keys(filter)[0];
      const filterType = Object.keys(filter[name])[0];
      const value = filter[name][filterType];

      return prevFilter[name]?.[filterType] !== value;
    });

    setQuery({ filters: { $and: nextFilters }, page: 1 });
  };

  return (
    query?.filters?.$and.map((filter, i) => {
      const attributeName = Object.keys(filter)[0];
      const attribute = filtersSchema.find(({ name }) => name === attributeName);

      if (!attribute) {
        return null;
      }

      if (attribute.fieldSchema.type === 'relation') {
        const relationTargetAttribute = attribute.fieldSchema.mainField.name;
        const filterObj = filter[attributeName][relationTargetAttribute];
        const operator = Object.keys(filterObj)[0];
        const value = filterObj[operator];

        return (
          <AttributeTag
            // eslint-disable-next-line react/no-array-index-key
            key={`${attributeName}-${i}`}
            attribute={attribute}
            filter={filter}
            onClick={handleClick}
            operator={operator}
            value={value}
          />
        );
      }

      const filterObj = filter[attributeName];
      const operator = Object.keys(filterObj)[0];
      const value = filterObj[operator];

      return (
        <AttributeTag
          // eslint-disable-next-line react/no-array-index-key
          key={`${attributeName}-${i}`}
          attribute={attribute}
          filter={filter}
          onClick={handleClick}
          operator={operator}
          value={value}
        />
      );
    }) || null
  );
};

FilterListURLQuery.defaultProps = {
  filtersSchema: [],
};

FilterListURLQuery.propTypes = {
  filtersSchema: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      metadatas: PropTypes.shape({ label: PropTypes.string }),
      fieldSchema: PropTypes.shape({
        type: PropTypes.string,
        mainField: PropTypes.shape({
          name: PropTypes.string,
          type: PropTypes.string,
        }),
      }),
    })
  ),
};

export default FilterListURLQuery;
