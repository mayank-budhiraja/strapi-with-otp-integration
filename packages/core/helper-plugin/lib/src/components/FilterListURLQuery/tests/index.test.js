/**
 *
 * Tests for FilterListURLQuery
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { ThemeProvider, lightTheme } from '@strapi/design-system';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import qs from 'qs';
import FilterListURLQuery from '../index';

const makeApp = (history, filtersSchema) => (
  <Router history={history}>
    <ThemeProvider theme={lightTheme}>
      <IntlProvider
        locale="en"
        messages={{
          'components.FilterOptions.FILTER_TYPES.$eq': 'is',
          'components.FilterOptions.FILTER_TYPES.$ne': 'is not',
          'components.FilterOptions.FILTER_TYPES.$contains': 'contains (case sensitive)',
          'components.FilterOptions.FILTER_TYPES.$notContains': 'does not contain (case sensitive)',
          'components.FilterOptions.FILTER_TYPES.$gt': 'is greater than',
          'components.FilterOptions.FILTER_TYPES.$gte': 'is greater than or equal to',
          'components.FilterOptions.FILTER_TYPES.$lt': 'is lower than',
          'components.FilterOptions.FILTER_TYPES.$lte': 'is lower than or equal to',
          'components.FilterOptions.FILTER_TYPES.$startsWith': 'starts with',
          'components.FilterOptions.FILTER_TYPES.$endsWith': 'ends with',
          'components.FilterOptions.FILTER_TYPES.$null': 'is null',
          'components.FilterOptions.FILTER_TYPES.$notNull': 'is not null',
        }}
        defaultLocale="en"
        textComponent="span"
      >
        <FilterListURLQuery filtersSchema={filtersSchema} />
      </IntlProvider>
    </ThemeProvider>
  </Router>
);

describe('<FilterListURLQuery />', () => {
  it('renders and matches the snapshot', () => {
    const history = createMemoryHistory();
    const filtersSchema = [
      {
        name: 'bool',
        metadatas: { label: 'Boolean' },
        fieldSchema: { type: 'boolean' },
      },
      {
        name: 'date',
        metadatas: { label: 'Date' },
        fieldSchema: { type: 'date' },
      },
      {
        name: 'enum',
        metadatas: { label: 'enum' },
        fieldSchema: { type: 'enumeration', options: ['one', 'two', 'three'] },
      },
      {
        name: 'long',
        metadatas: { label: 'Long' },
        fieldSchema: { type: 'string' },
      },
      {
        name: 'city',
        metadatas: { label: 'city' },
        fieldSchema: { type: 'string' },
      },
      {
        name: 'country',
        metadatas: { label: 'country' },
        fieldSchema: { type: 'string' },
      },
      {
        name: 'many_to_one',
        metadatas: { label: 'many to one' },
        fieldSchema: {
          type: 'relation',
          mainField: {
            name: 'postal_code',
            schema: { type: 'string', pluginOptions: { i18n: { localized: true } } },
          },
        },
      },
      {
        name: 'number',
        metadatas: { label: 'number' },
        fieldSchema: { type: 'integer' },
      },
      {
        name: 'float',
        metadatas: { label: 'Float' },
        fieldSchema: { type: 'float' },
      },
      {
        name: 'time',
        metadatas: { label: 'time' },
        fieldSchema: { type: 'time' },
      },
    ];

    const search = {
      sort: 'city',
      filters: {
        $and: [
          { bool: { $eq: 'true' } },
          { date: { $ne: '2021-09-01' } },
          { city: { $null: 'true' } },
          { country: { $notNull: 'true' } },
          { time: { $contains: '00:45' } },
          { many_to_one: { postal_code: { $notContains: 'test' } } },
          { many_to_one: { postal_code: { $eq: 'test' } } },
          { city: { $startsWith: 'paris' } },
          { country: { $endsWith: 'france' } },
          { number: { $gt: '2' } },
          { float: { $gte: '1' } },
          { float: { $lte: '3' } },
          { float: { $lt: '4' } },
        ],
      },
    };
    history.push({
      pathname: '/',
      search: qs.stringify(search, { encode: false }),
    });

    const { container } = render(makeApp(history, filtersSchema));

    expect(container).toMatchSnapshot();
  });
});
