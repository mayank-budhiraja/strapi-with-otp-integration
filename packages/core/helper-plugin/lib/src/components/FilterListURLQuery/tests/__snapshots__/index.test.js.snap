// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterListURLQuery /> renders and matches the snapshot 1`] = `
.c7 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  padding: 4px;
}

.c4 {
  font-weight: 600;
  color: #32324d;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c1 {
  background: #f0f0ff;
  color: #4945ff;
  padding-right: 12px;
  padding-left: 12px;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  border-color: #d9d8ff;
  height: 2rem;
}

.c6 {
  padding-left: 8px;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 svg {
  height: 0.5rem;
  width: 0.5rem;
}

.c2 svg path {
  fill: #4945ff;
}

.c5 {
  border-right: 1px solid #d9d8ff;
  color: inherit;
  padding-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          Boolean is true
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          Date is not Wednesday, September 1, 2021
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          city is null 
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          country is not null 
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          time contains (case sensitive) 12:45 AM
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          many to one does not contain (case sensitive) test
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          many to one is test
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          city starts with paris
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          country ends with france
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          number is greater than 2
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          Float is greater than or equal to 1
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          Float is lower than or equal to 3
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      class="c1 c2"
      height="2rem"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          Float is lower than 4
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#212134"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c7"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
